pkgs/net-doc/net/scribblings/server.scrbl

```diff
@@ -0,0 +1,176 @@
+                                      (-> (integer-in 0 65535)
+                                          exact-nonnegative-integer?
+                                          boolean?
+                                          string?
+                                          (and/c listener? evt?))
+                                      tcp-listen]
+                       [#:accept-proc accept
+                                      (-> listener? (values input-port? output-port?))
+                                      tcp-accept]
+                       [#:close-proc close
+                                     (-> listener? void?)
+                                     tcp-close]
+                       [#:timeout-evt-proc make-timeout-evt
+                                           (-> thread? input-port? output-port? boolean? evt?)
+                                           (λ (thd in out break-sent?) never-evt)]) (-> void?)]{
+
+  Creates a listening server on @racket[port], bound to the interface
+  associated with @racket[host].  After the listener is ready, it
+  spawns a background thread to accept new connections.  For every new
+  connection, @racket[handle] is called with two arguments: an input
+  port to read from the client, and an output port to write to the
+  client.  The returned procedure stops the accepting thread and
+  closes the listener but does not terminate active connections.
+
+  Each client connection is managed by a new custodian, and each call
+  to @racket[handle] occurs in a new thread (also managed by the
+  connection's custodian).  Each handling thread has an associated
+  supervising thread that shuts down the connection's custodian when
+  the handling thread terminates or when the result of
+  @racket[make-timeout-evt] is ready for synchronization.  Breaks are
+  enabled in handling threads if breaks are enabled when
+  @racket[start-server] is called.  Handling threads need not close
+  the provided input and output ports.
+
+  The @racket[make-timeout-evt] procedure is called with the
+  connection-handling thread, the input port and the output port of
+  every new connection, and a boolean to signal if the handling thread
+  has already been sent a break.  When the event it returns is ready
+  for synchronization and if the handling thread is still running, the
+  handling thread is sent a break and @racket[make-timeout-evt] is
+  called again to produce an event that, when ready for
+  synchronization, will cause the connection's custodian to be shut
+  down and, consequently, the handling thread to be killed if it is
+  still running by that time.
+
+  The server keeps track of the number of active connections and
+  pauses accepting new connections once that number reaches
+  @racket[max-concurrent], resuming once the number goes down again.
+
+  The @racket[reuse?] and @racket[max-allow-wait] arguments have the
+  same meaning as in @racket[tcp-listen].  The @racket[listen],
```

Comment:
I'm of two minds about these: handling them as arguments to `start-server` is obviously much more ergonomic for TCP, but I wonder how well they generalize to other kinds of connections.

Reply:
Actually, as I look more closely, it seems `port`, `max-allow-wait`, `reuse?`, and `host` are all used only by passing them to `listen` exactly once, before starting threads or looping or any of the other complexity:
```racket
  (define listener
    (listen port max-allow-wait reuse? host))
```
What would you think of dropping all of them and having `start-server` take the `listener` as an argument, instead?

Reply:
Thanks for all the feedback so far!

> What would you think of dropping all of them and having start-server take the listener as an argument, instead? 

I'm reluctant to make that change, because it feels wrong for the listener to be created externally to `start-server` and  for `start-server` to then take ownership of the listener and be in charge of closing it. My preference would be to reduce the set of arguments that `start-server` takes and rely on the caller to provide a customized `listen-proc`.

Reply:
After working on the new set of examples, I think it's probably worth keeping the keywords as-is. Partly because TCP is likely to be the common use case, but also because of the implementation of `tcp-listen`: the `host` argument is the final argument, so we'd have to provide our own defaults to `listen-proc` anyway (which is what `mzlib`'s `run-server` does, but seems worse). Having the `host` argument be a required arg to `start-server` seems better than relying on `tcp-listen`'s default of listening on all interfaces, too.